### PR TITLE
fix(cairo): o3 copy to use index block capacity

### PR DESCRIPTION
**Pull request by ideoma**
When Out of Order insert happens and index is re-written the index file is initialised with minimum Index Block Capacity equals to 4. This kills performance of Out of Order index rewrite on the partition since block capacity of 4 forces to add a new block for every 4 values. 

It is especially painful when there are few distinct values of the column with hundred thousands of values.

The fix makes sure that when index is re-written during O3 insert the most up to date value of Index Block Capacity is used from table metadata

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 17 / 17 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3OpenColumnTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3OpenColumnJob.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 17 / 17 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3OpenColumnTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3OpenColumnJob.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 17 / 17 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3OpenColumnTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3OpenColumnJob.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|3|3|100.00%|

File: core/src/main/java/io/questdb/cairo/O3PartitionJob.java [hunk 1]
```diff
@@ -783,15 +783,16 @@ private static void publishOpenColumnTasks(
                 }
 
                 final CharSequence columnName = metadata.getColumnName(i);
-                final boolean isIndexed = metadata.isColumnIndexed(i);
+                final boolean indexBlockCapacity = metadata.isColumnIndexed(i);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/O3PartitionJob.java:786]: confusing naming

File: core/src/main/java/io/questdb/cairo/O3PartitionJob.java [hunk 2]
```diff
@@ -783,15 +783,16 @@ private static void publishOpenColumnTasks(
...
+                final int indexValueCapacity = indexBlockCapacity ? metadata.getIndexValueBlockCapacity(i) : -1;
                 if (openColumnMode == OPEN_LAST_PARTITION_FOR_APPEND || openColumnMode == OPEN_LAST_PARTITION_FOR_MERGE) {
                     srcDataTop = tableWriter.getColumnTop(i);
                 } else {
                     srcDataTop = -1; // column open job will have to find out if top exists and its value
                 }
 
                 final BitmapIndexWriter indexWriter;
-                if (isIndexed) {
+                if (indexBlockCapacity) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/O3PartitionJob.java:795]: same

File: core/src/main/java/io/questdb/cairo/O3PartitionJob.java [hunk 3]
```diff
@@ -833,7 +834,7 @@ private static void publishOpenColumnTasks(
                                 suffixType,
                                 suffixLo,
                                 suffixHi,
-                                isIndexed,
+                                indexValueCapacity,
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/O3PartitionJob.java:837]: inconsistent parameter name

File: core/src/main/java/io/questdb/cairo/O3CopyJob.java [hunk 4]
```diff
@@ -469,14 +470,15 @@ private static void updateIndex(
             long srcTimestampAddr,
             long srcTimestampSize,
             TableWriter tableWriter,
-            BitmapIndexWriter indexWriter
+            BitmapIndexWriter indexWriter,
+            int indexValueCapacity
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/O3CopyJob.java:474]: inconsistent parameter name

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by bluestreak01** [comment]
it would also be good to harmonise `initKeyMemory` parameter name with variable names that are passed to that parameter

File: core/src/main/java/io/questdb/cairo/O3PartitionJob.java (hunk 2 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/O3PartitionJob.java:795]: ok

File: core/src/main/java/io/questdb/cairo/O3PartitionJob.java (hunk 3 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/O3PartitionJob.java:837]: ok

File: core/src/main/java/io/questdb/cairo/O3CopyJob.java (hunk 4 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/O3CopyJob.java:474]: ok

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 16 / 16 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3OpenColumnTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3OpenColumnJob.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 16 / 16 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3OpenColumnTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3OpenColumnJob.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 23 / 23 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3OpenColumnTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3OpenColumnJob.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 23 / 23 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3OpenColumnTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3OpenColumnJob.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 23 / 23 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3OpenColumnTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3CopyJob.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3OpenColumnJob.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CopyTask.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|1|1|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**